src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -830,34 +840,34 @@ internal bool MethodGroupIsCompatibleWithDelegate(BoundExpression? receiverOpt,
 
             for (int i = 0; i < numParams; i++)
             {
-                var delegateParameter = delegateParameters[i];
+                var delegateParameter = delegateOrFuncPtrParameters[i];
                 var methodParameter = methodParameters[isExtensionMethod ? i + 1 : i];
 
                 if (delegateParameter.RefKind != methodParameter.RefKind ||
-                    !Conversions.HasIdentityOrImplicitReferenceConversion(delegateParameter.Type, methodParameter.Type, ref useSiteDiagnostics))
+                    !hasVariantConversion(delegateType.TypeKind, Conversions, delegateParameter.Type, methodParameter.Type, ref useSiteDiagnostics))
```

Comment:
>hasVariantConversion [](start = 21, length = 20)

You should not permit a variant conversion when there is a ref kind other than none.

Reply:
I'll try to dig into this more. Currently, I'm not changing the existing semantics of the code, which seems incorrect for delegate conversions today. However, I couldn't find a scenario where this actually manifested in illegal code being allowed. Will keep digging.

Reply:
It might make sense to find where the error is reported for that scenario, and add a comment here that the scenario is handled *wherever*.

---
In reply to: [393951430](https://github.com/dotnet/roslyn/pull/42488#discussion_r393951430) [](ancestors = 393951430)

Reply:
Alright, I've written up a detailed comment as to why this check exists, and I've also opened a C# proposal to actually change the behavior here: https://github.com/dotnet/csharplang/issues/3277